python/tk_framework_desktopserver/shotgun/scripts/cache_commands.py

```diff
@@ -0,0 +1,185 @@
+    sgtk.LogManager().initialize_base_file_handler("tk-shotgun")
+    sgtk_logger.debug("Log dir: %s" % (sgtk.LogManager().log_folder))
+
+    return sgtk_logger, bootstrap_log_handler
+
+def cache(cache_file, data, base_configuration, hash_data, engine_name):
+    import sqlite3
+    import sgtk
+
+    entity = dict(type=data["entity_type"], id=data["entity_id"])
+    project_entity = dict(id=data["project_id"], type="Project")
+
+    # Setup the bootstrap manager.
+    toolkit_mgr = sgtk.bootstrap.ToolkitManager()
+    toolkit_mgr.plugin_id = "shotgun_toolkit_menu_caching"
+    toolkit_mgr.base_configuration = base_configuration
+
+    for pc in data["pipeline_configs"]:
+        pc["project"] = pc.get("project", project_entity)
+    import pprint
+    pprint.pprint(data["pipeline_configs"])
+    pcs = toolkit_mgr.sort_and_filter_configuration_entities(
+        project=project_entity,
+        entities=data["pipeline_configs"],
+    ) or [dict(id=None)]
+
+    for pc in pcs:
+        logger, log_handler = get_sgtk_logger(sgtk)
+        pprint.pprint(hash_data)
+        lookup_hash = hash_data[pc["id"]]["lookup_hash"]
+        contents_hash = hash_data[pc["id"]]["contents_hash"]
+
+        toolkit_mgr.pipeline_configuration = pc["id"]
+        engine = toolkit_mgr.bootstrap_engine(engine_name, entity=entity)
+        pc_descriptor = toolkit_mgr.resolve_descriptor(project_entity)
+
+        # Clean up the pre-bootstrap logger since we can use the engine's
+        # logger from here on out.
+        sgtk.LogManager().root_logger.removeHandler(log_handler)
+        engine.logger.debug("Processing engine commands...")
+        commands = []
+
+        for cmd_name, data in engine.commands.iteritems():
+            engine.logger.debug("Processing command: %s" % cmd_name)
+            props = data["properties"]
+            app = props.get("app")
+
+            if app:
+                app_name = app.name
+            else:
+                app_name = None
+
+            commands.append(
+                dict(
+                    name=cmd_name,
+                    title=props.get("title", cmd_name),
+                    deny_permissions=props.get("deny_permissions", []),
+                    supports_multiple_selection=props.get(
+                        "supports_multiple_selection",
+                        False
+                    ),
+                    app_name=app_name,
+                    group=props.get("group"),
+                    group_default=props.get("group_default"),
+                    engine_name=props.get("engine_name"),
+                ),
+            )
+
+        engine.logger.debug("Engine commands processed.")
+        engine.logger.debug("Inserting commands into cache...")
+
+        # Connect to the database and get the hashes we need to include in
+        # the insert. Each of the lookups call out to the browser_integration
+        # core hook.
+        connection = sqlite3.connect(cache_file)
```

Comment:
`with sqlite3.connect(cache_file) as connection:` would yield code without `try: except`. I think you won't need to call commit either.